text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+`std::module_path` and expands to an expression of type `&'static str`.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently macros for getting the source file name, `std::file`, and
+the line number, `std::line`. To complete this pair, a macro to for getting
+the current function name would be useful. All three macros are useful for
+e.g. generating log output with more context about where something was
+happening. The file and line is good for being able to look up the exact
+location, however the fully qualified function name gives additional context
+without even having to look into the source code.
+
+# Detailed design
+[design]: #detailed-design
+
+This feature would have an equivalent design and implementation to
+`std::file`. It would return a fully qualified function name but the actual
+string returned is implementation defined. The string is meant to be used by
+the user to get an idea about the context of the code in e.g. debug output,
+and should be meaningful for a human, but is not meant to be analyzed by a
+program.
+
+For example the following scheme could be used:
+
+ - Plain functions use their name, e.g. `some_function` for
+
+````
+    fn some_function(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Generic functions use their name, followed by the generic arguments, e.g.
+   `some_function<T, U>` for
+
+````
+    fn some_function<T, U>(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Functions inside non-generic `impl` blocks for structs would use the name of the
+   struct, followed by `::` and the function name, e.g.
+   `SomeStruct::some_function` for
+
+````
+    struct SomeStruct {...}
+
+    impl SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside generic `impl` blocks for structs would use the name of the
+   struct, followed by the generic parameters as given by the `impl` block,
+   followed by `::` and the function name, e.g.
+   `SomeStruct<i32>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl SomeStruct<i32> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Similar for generic `impl` blocks, e.g. `SomeStruct<T>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl<T> SomeStruct<T> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside trait `impl` blocks would use, enclosed in angle brackets,
+   the name of the type, followed by `as`, followed by the trait name, and
+   then `::` followed by the name of the function, e.g.
+   `<SomeStruct as SomeTrait>::some_function` for
+
+````
+    impl SomeTrait for SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Closures are using the name of the surrounding function, followed by `::`
+   and the string `<closure/definition_line_number>`, where `definition_line_number` is the number
```

Comment:
It's a bit ugly, however you otherwise won't have any way to distinguish two closures that are defined inside the same function. Instead of the line number, we could also just count though. Unless someone has an idea how to easily give a unique name to closures